Added Location.create

pgdr opened the pull request:
To assist end users in creating location and setting location to a certain device, we want:

`api/create_location/` with post `id=FriskPiXY, key=000123..., latitude=66.71, longitude=5.5, altitude=50, name=Minde`

to construct a new location with aforementioned values, and set device id `FriskPiXY` to have this location, give, of course, that device `FriskPiXY` is associated with the provided API key.

Goal: #205

semafor commented on sensor/models.py:
```diff
@@ -29,6 +29,42 @@ class Location(Model):
+        try:
+            device = Device.objects.get( pk = dev_id )
+        except Device.DoesNotExist:
+            raise ValueError('No such device "%s".' % dev_id)
```
keyerror? :)

semafor commented on sensor/models.py:
```diff
@@ -29,6 +29,42 @@ class Location(Model):
+            raise KeyError('Needs %s to create location!' % (err if err else dict_key))
+
+    @classmethod
+    def create(cls, data):
```
LGTM!

joakim-hove commented on sensor/models.py:
```diff
@@ -29,6 +29,42 @@ class Location(Model):
+        _assert_field_for_create('key', data, err='api key')
+        _assert_field_for_create('latitude', data)
+        _assert_field_for_create('longitude', data)
+        _assert_field_for_create('altitude', data)
```
The `altitude` field is currently not required in the `Location` model. If we make it required in this input validation I think we should start changing the `Location`model to make the `altitude` field required.

joakim-hove commented:
Small PR - large comments:

1. Looking at the new `create( )` method makes me slightly uneasy: I have for a long time wanted to start using the [Django Rest Framework](http://www.django-rest-framework.org/) - but not really got it going. If you agree with me I really feel the `create( )` method is a prime candidate for a REST serializer?

2. I feel the verification of device ID should be done in view code prior to calling the `create( )` method?

joakim-hove commented on sensor/models.py:
```diff
@@ -29,6 +29,42 @@ class Location(Model):
+            lon = '%.8f' % float(lon)
+            alt = '%d' % int(float(alt))
+        except ValueError as err:
+            raise ValueError('Latitude, longitude, and altitude must be numbers. %s' % err)
```
This should be handled by the Django layer, no need to do manual numerical conversions.

pgdr commented:
@joakim-hove @semafor Thanks for valuable feedback.  I will implement all the requested changes.

pgdr commented:
@joakim-hove this is a "working" example of the functionality we need in order to "complete" the controlpanel.

It allows us to send the URL
```url
sensor/view/location/create/FriskPiId/?key=xyz
                                      &longitude=60.0&latitude=5.0
                                      &altitude=70&name=My new place
```
to create a new location for `FriskPiId` provided the key is correct.

Now, there are a lot of things we can improve, and there is no testing, but I would like this merged _sooner_ rather than _later_.  Hence, I aim for minimal requirements to get your approval.

joakim-hove commented:
Ok - please merge it.

pgdr commented:
Thanks, but I think that I first want to see if I can add a couple of tests and also make _altitude_ optional.

In addition I have two questions:
1. You mentioned that I shouldn't use `Location.save` for creating a new location, but rather `Location.objects.create_location`, is that right?
2. I would like to _redirect_ to the `device` view after the location is created (so that the URL is also changed), i.e., to the site `sensor/view/device/FriskPIid/`.  Do you have an idea of how to do that?

joakim-hove commented:
1. Location.objects.create()
2. from django.http import HttpResponseRedirect

pgdr commented:
I addressed those two, and I moved the authentication code out to the View.

I still have a goal of adding a test before merging.

Are there more issues that you can see that obviously _should be fixed_ (except using the Django Rest Framework)?

joakim-hove commented:
1. The create method will save - no need to call loc.save()
2. No reason to check the fields, and then raise KeyError - just access the fields in the create call - and let Python give you the KeyError.
3. Testing with the Django test client is quite simple.

joakim-hove commented on sensor/view/views.py:
```diff
@@ -17,9 +26,33 @@ def get(self , request):
+
+class LocationCreator(View):
+
+    def get(self, request, pk):
```
This should be POST?

joakim-hove commented on sensor/view/urls.py:
```diff
@@ -7,8 +7,7 @@
     url(r'^$' , MainView.as_view(), name = "sensor.view.main"),
     url(r'^device/(?P<pk>%s)/$' % models.Device.IDPattern, DeviceView.as_view() , name = "view.device.info"),
+    url(r'^location/create/(?P<pk>%s)/$' % models.Device.IDPattern,
+        LocationCreator.as_view(), name="view.location.creator"),
```
This should be in api_urls.py

pgdr commented:
Thanks a lot, I appreciate it.  I will implement your suggestions and a test and then I'll merge it.

pgdr commented:
@joakim-hove @semafor Okay, thorough tests are in place.  I'll merge this once Travis greenlights.

joakim-hove commented on sensor/api/api_views.py:
```diff
@@ -127,21 +135,46 @@ def put(self , request , *args, **kwargs):
+        loc = Location.create(payload)
+        device.location = loc
+        device.save()
+        return HttpResponseRedirect('/sensor/view/device/%s/' % device.id)
```
Try to use the `reverse` functionality to look up a named url here, that way the url structure is *only* configured in the `urls.py` scheme - and not scattered around in the code.

joakim-hove commented on sensor/tests/test_location.py:
```diff
@@ -0,0 +1,113 @@
+    def test_create_location_missing_fields(self):
+        with self.assertRaises(KeyError) as ctx:  # missing latitude
+            args = (self.url, self.dev_id, self.key, self.lon, self.name)
+            response = self.client.get("%s/%s/?key=%s&longitude=%s&name=%s" %
```
1. Use the [`reverse`](https://docs.djangoproject.com/en/1.11/ref/urlresolvers/) functionality instead of hardcoding the url structure.
2. Instead of passing `?key1=value1&key2=value2` explcitly pass a dictinary as the optional `data` argument:
```
client.get( url, param,   data = {"key" : key_variable, "longitude" : ...})
```

pgdr commented:
Comments addressed:
1. use `reverse` in `api_urls`
2.  use `reverse` in tests
3. use `client.get(url, data=data)`  in tests
will fix commits now

pgdr commented:
Thanks for all the feedback, @joakim-hove ... the quality has improved **immensly** since the first attempt.
